// pirsch-sdk/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PirschEndpoint = exports.PIRSCH_URL_LENGTH_LIMIT = exports.PIRSCH_IDENTIFICATION_CODE_LENGTH = exports.PIRSCH_ACCESS_TOKEN_LENGTH = exports.PIRSCH_CLIENT_SECRET_LENGTH = exports.PIRSCH_CLIENT_ID_LENGTH = exports.PIRSCH_ACCESS_TOKEN_PREFIX = exports.PIRSCH_PROXY_HEADERS = exports.PIRSCH_REFERRER_QUERY_PARAMETERS = exports.PIRSCH_DEFAULT_PROTOCOL = exports.PIRSCH_DEFAULT_TIMEOUT = exports.PIRSCH_DEFAULT_BASE_URL = void 0;
exports.PIRSCH_DEFAULT_BASE_URL = "https://api.pirsch.io";
exports.PIRSCH_DEFAULT_TIMEOUT = 5000;
exports.PIRSCH_DEFAULT_PROTOCOL = "https";
exports.PIRSCH_REFERRER_QUERY_PARAMETERS = ["ref", "referer", "referrer", "source", "utm_source"];
exports.PIRSCH_PROXY_HEADERS = ["cf-connecting-ip", "x-forwarded-for", "forwarded", "x-real-ip"];
exports.PIRSCH_ACCESS_TOKEN_PREFIX = "pa_";
exports.PIRSCH_CLIENT_ID_LENGTH = 32;
exports.PIRSCH_CLIENT_SECRET_LENGTH = 64;
exports.PIRSCH_ACCESS_TOKEN_LENGTH = 45;
exports.PIRSCH_IDENTIFICATION_CODE_LENGTH = 32;
exports.PIRSCH_URL_LENGTH_LIMIT = 1800;
var PirschEndpoint;
(function (PirschEndpoint) {
    PirschEndpoint["AUTHENTICATION"] = "token";
    PirschEndpoint["HIT"] = "hit";
    PirschEndpoint["HIT_BATCH"] = "hit/batch";
    PirschEndpoint["EVENT"] = "event";
    PirschEndpoint["EVENT_BATCH"] = "event/batch";
    PirschEndpoint["SESSION"] = "session";
    PirschEndpoint["SESSION_BATCH"] = "session/batch";
    PirschEndpoint["DOMAIN"] = "domain";
    PirschEndpoint["SESSION_DURATION"] = "statistics/duration/session";
    PirschEndpoint["TIME_ON_PAGE"] = "statistics/duration/page";
    PirschEndpoint["UTM_SOURCE"] = "statistics/utm/source";
    PirschEndpoint["UTM_MEDIUM"] = "statistics/utm/medium";
    PirschEndpoint["UTM_CAMPAIGN"] = "statistics/utm/campaign";
    PirschEndpoint["UTM_CONTENT"] = "statistics/utm/content";
    PirschEndpoint["UTM_TERM"] = "statistics/utm/term";
    PirschEndpoint["TOTAL_VISITORS"] = "statistics/total";
    PirschEndpoint["VISITORS"] = "statistics/visitor";
    PirschEndpoint["PAGES"] = "statistics/page";
    PirschEndpoint["ENTRY_PAGES"] = "statistics/page/entry";
    PirschEndpoint["EXIT_PAGES"] = "statistics/page/exit";
    PirschEndpoint["CONVERSION_GOALS"] = "statistics/goals";
    PirschEndpoint["EVENTS"] = "statistics/events";
    PirschEndpoint["EVENT_METADATA"] = "statistics/event/meta";
    PirschEndpoint["LIST_EVENTS"] = "statistics/event/list";
    PirschEndpoint["GROWTH_RATE"] = "statistics/growth";
    PirschEndpoint["ACTIVE_VISITORS"] = "statistics/active";
    PirschEndpoint["TIME_OF_DAY"] = "statistics/hours";
    PirschEndpoint["LANGUAGE"] = "statistics/language";
    PirschEndpoint["REFERRER"] = "statistics/referrer";
    PirschEndpoint["OS"] = "statistics/os";
    PirschEndpoint["OS_VERSION"] = "statistics/os/version";
    PirschEndpoint["BROWSER"] = "statistics/browser";
    PirschEndpoint["BROWSER_VERSION"] = "statistics/browser/version";
    PirschEndpoint["COUNTRY"] = "statistics/country";
    PirschEndpoint["CITY"] = "statistics/city";
    PirschEndpoint["PLATFORM"] = "statistics/platform";
    PirschEndpoint["SCREEN"] = "statistics/screen";
    PirschEndpoint["KEYWORDS"] = "statistics/keywords";
})(PirschEndpoint = exports.PirschEndpoint || (exports.PirschEndpoint = {}));
